File: 20_杂题/atc競プロ/AtCoder Beginner Contest/278/f.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

// from functools import lru_cache
// import sys

// sys.setrecursionlimit(int(1e9))
// input = lambda: sys.stdin.readline().rstrip("\r\n")
// MOD = 998244353
// INF = int(4e18)
// # N 個の文字列 S
// # 1
// # ​
// #  ,S
// # 2
// # ​
// #  ,…,S
// # N
// # ​
// #   が与えられます。 S
// # i
// # ​
// #   (1≤i≤N) は英小文字からなる長さ 10 以下の空でない文字列で、互いに異なります。

// # 先手太郎君と後手次郎君がしりとりをします。 このしりとりでは、先手太郎君と後手次郎君の手番が交互に訪れます。 はじめの手番は先手太郎君の手番です。 それぞれのプレイヤーは自分の手番において整数 i (1≤i≤N) を 1 つ選びます。 このとき、i は次の 2 つの条件を満たしていなければなりません。

// # i は、しりとりが開始してからこれまでの 2 人の手番で選ばれたどの整数とも異なる
// # この手番がしりとりの最初の手番であるか、直前に選ばれた整数を j として、S
// # j
// # ​
// #   の最後の文字と S
// # i
// # ​
// #   の最初の文字が等しい
// # 条件を満たす i を選べなくなったプレイヤーの負けで、負けなかったプレイヤーの勝ちです。

// # 2 人が最適に行動したときに勝つのはどちらかを判定してください。
// if __name__ == "__main__":
//     n = int(input())
//     words = [input() for _ in range(n)]

//     @lru_cache(None)
//     def dfs(visited: int, pre: str) -> bool:
//         for i in range(n):
//             if visited & (1 << i) or words[i][0] != pre:
//                 continue
//             if not dfs(visited | (1 << i), words[i][-1]):
//                 return True
//         return False

//     for start in range(n):
//         if dfs(1 << start, words[start][0]):
//             print("First")
//             exit(0)
//     print("Second")

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n int
	fmt.Fscan(in, &n)
	words := make([]string, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &words[i])
	}

	N := (1 << n) * 32
	memo := make([]int, N)
	for i := 0; i < N; i++ {
		memo[i] = -1
	}

	var dfs func(visited int, pre byte) int
	dfs = func(visited int, pre byte) int {
		hash := visited*30 + int(pre)
		if memo[hash] != -1 {
			return memo[hash]
		}

		for i := 0; i < n; i++ {
			if visited&(1<<i) == 0 && (pre == 27 || words[i][0]-'a' == pre) {
				if dfs(visited|(1<<i), words[i][len(words[i])-1]-'a') == 0 {
					memo[hash] = 1
					return 1
				}
			}
		}
		memo[hash] = 0
		return 0
	}

	ok := false
	for start := 0; start < n; start++ {
		if dfs(0, 27) == 1 {
			fmt.Fprintln(out, "First")
			ok = true
			break
		}
	}

	if !ok {
		fmt.Fprintln(out, "Second")
	}
}

File: leetcode/abc333/f.py
import sys

sys.setrecursionlimit(int(1e6))
input = lambda: sys.stdin.readline().rstrip("\r\n")
MOD = 998244353
INF = int(4e18)
# N 人の人が一列に並んでおり、人
# i は先頭から
# i 番目に並んでいます。

# 以下の操作を、列に並んでいる人が
# 1 人になるまで繰り返します。

# 先頭に並んでいる人を
# 2
# 1
# ​
#   の確率で列から取り除き、そうでない場合は列の末尾に移す。
# 人
# i=1,2,…,N それぞれについて、人
# i が最後まで列に並んでいる
# 1 人になる確率を
# mod 998244353 で求めて下さい。(取り除くかどうかの選択はすべてランダムかつ独立です。)

INV2 = (MOD + 1) // 2
if __name__ == "__main__":
    n = int(input())
    # void rd() {
    #   cin >> n >> m;
    # }

    # cin>>n>>m;
    # f[1][1]=1;
    # for(int i=2;i<=n;i++){
    # 	double s=0.5,t=0;
    # 	for(int j=2;j<=i;j++){
    # 		s=s/2;
    # 		t=t/2+f[i-1][j-1]/3;
    # 	}
    # 	f[i][i]=t/(1-s);
    # 	f[i][1]=f[i][i]/2;
    # 	for(int j=2;j<i;j++)
    # 		f[i][j]=f[i][j-1]/2+f[i-1][j-1]/2;
    # }printf("%.9lf",f[n][m]);

    dp = [[0] * (n + 1) for _ in range(n + 1)]
    dp[1][1] = 1
    for i in range(2, n + 1):
        dp[i][1] = 1
        s = INV2
        for j in range(2, i + 1):
            s = s * INV2 % MOD
            dp[i][j] = (dp[i - 1][j - 1] * INV2 + dp[i][j - 1] * INV2) % MOD
        dp[i][i] = dp[i][i - 1] * INV2 % MOD
        for j in range(2, i):
            dp[i][j] = (dp[i][j - 1] * INV2 + dp[i - 1][j - 1] * INV2) % MOD

    for i in range(1, n + 1):
        print(dp[n][i], end=" ")
